src/coreclr/src/vm/processdiagnosticsprotocolhelper.cpp

```diff
@@ -114,6 +136,146 @@ bool ProcessInfoPayload::Flatten(BYTE * &lpBuffer, uint16_t &cbSize)
     return fSuccess;
 }
 
+void EnvironmentHelper::PopulateEnvironment()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    if (Environment != nullptr)
+        return;
+
+    // env block is an array of strings of the form "key=value" delimited by null and terminated by null
+    // e.g., "key=value\0key=value\0\0";
+    LPWSTR envBlock = GetEnvironmentStringsW();
+    LPWSTR envCursor = envBlock;
+    // first calculate the buffer size
+    uint32_t nWchars = 0;
+    uint32_t nEntries = 0;
+    while(*envCursor != 0) 
+    {
+        uint32_t len = static_cast<uint32_t>(wcslen(envCursor) + 1);
+        nEntries++;
+        nWchars += len;
+        envCursor += len;
+    }
+
+    // serialized size value can't be larger than uint32_t
+    S_UINT32 size(sizeof(uint32_t) + (nEntries * sizeof(uint32_t)) + (nWchars * sizeof(WCHAR)));
+    // if the envblock size value is larger than a uint32_t then we should bail
+    if (!size.IsOverflow())
+    {
+        // copy out the Environment block
+        LPWSTR tmpEnv = new (nothrow) WCHAR[nWchars + 1];
+        if (tmpEnv != nullptr)
+        {
+            envCursor = envBlock;
+            LPWSTR payloadCursor = tmpEnv;
+
+            while(*envCursor != 0) 
```

Comment:
Rather than iterating the environment block again, could you track the overall size of the block when you calculate the number of entries and characters above and then just do a bulk copy of the entire block here?